File: 2021/day01/main.hs
solve :: [Int] -> Int
solve xs = length . filter (uncurry (<)) $ zip xs (tail xs)

window :: Int -> [Int] -> [[Int]]
window n [] = []
window n xs = (take n xs) : window n (tail xs)

main :: IO()
main = do
  nums <- map read.lines <$> (readFile "input") :: IO [Int]
  print $ solve nums
  print $ (solve . map sum . window 3) nums

File: 2021/day01/main.clj
(require '[clojure.string :as str])

(def nums (->> "input"
               slurp
               str/trim
               str/split-lines
               (map #(Integer/parseInt %))))

(defn window [n nums]
  (loop [xs nums ret '()]
    (if (empty? xs)
      (reverse ret)
      (recur (rest xs) (cons (take n xs) ret)))))

(defn silver [nums]
  (->> (map vector nums (drop 1 nums))
       (filter (partial apply <))
       count))

(def gold
  (->> nums
       (window 3)
       (map #(reduce + %))
       silver))

(println (silver nums) gold)